crates/fuel-core/src/coins_query.rs

```diff
@@ -221,6 +274,168 @@ pub async fn random_improve(
+        dust_coins_stream,
+        max_dust_count,
+        excluded_ids,
+        |coin, _| coin == last_big_coin,
+    )
+    .await
+}
+
+async fn select_coins_until<F>(
+    mut coins_stream: impl Stream<Item = StorageResult<CoinsToSpendIndexEntry>> + Unpin,
+    max: u16,
+    excluded_ids: &ExcludedCoinIds<'_>,
+    predicate: F,
+) -> Result<(u64, Vec<CoinsToSpendIndexEntry>), CoinsQueryError>
+where
+    F: Fn(&CoinsToSpendIndexEntry, u64) -> bool,
+{
+    let mut coins_total_value: u64 = 0;
+    let mut count = 0;
+    let mut coins = Vec::with_capacity(max as usize);
+    while let Some(coin) = coins_stream.next().await {
+        let coin = coin?;
+        if !is_excluded(&coin, excluded_ids)? {
+            if count >= max || predicate(&coin, coins_total_value) {
+                break;
+            }
+            count = count.saturating_add(1);
+            let amount = coin.0.amount();
+            coins_total_value = coins_total_value.saturating_add(amount);
+            coins.push(coin);
+        }
+    }
+    Ok((coins_total_value, coins))
+}
+
+fn is_excluded(
+    (key, coin_type): &CoinsToSpendIndexEntry,
+    excluded_ids: &ExcludedCoinIds,
+) -> Result<bool, CoinsQueryError> {
+    match coin_type {
+        IndexedCoinType::Coin => {
+            let utxo_id_bytes = key.foreign_key_bytes();
+            let tx_id: TxId = utxo_id_bytes
+                .get(..32)
+                .ok_or(CoinsQueryError::IncorrectCoinKeyInIndex)?
+                .try_into()
+                .map_err(|_| CoinsQueryError::IncorrectCoinKeyInIndex)?;
+
+            let output_index = u16::from_be_bytes(
+                utxo_id_bytes
+                    .get(32..34)
+                    .ok_or(CoinsQueryError::IncorrectCoinKeyInIndex)?
+                    .try_into()
+                    .map_err(|_| CoinsQueryError::IncorrectCoinKeyInIndex)?,
+            );
+            Ok(excluded_ids.is_coin_excluded(&UtxoId::new(tx_id, output_index)))
+        }
+        IndexedCoinType::Message => {
+            let nonce_bytes = key.foreign_key_bytes();
+            let nonce: Nonce = nonce_bytes
+                .get(..)
+                .ok_or(CoinsQueryError::IncorrectMessageKeyInIndex)?
+                .try_into()
+                .map_err(|_| CoinsQueryError::IncorrectMessageKeyInIndex)?;
+            Ok(excluded_ids.is_message_excluded(&nonce))
+        }
+    }
+}
+
+fn max_dust_count(max: u16, big_coins_len: u16) -> u16 {
+    let mut rng = rand::thread_rng();
+    rng.gen_range(0..=max.saturating_sub(big_coins_len))
+}
+
+fn skip_big_coins_up_to_amount(
+    big_coins: impl IntoIterator<Item = CoinsToSpendIndexEntry>,
+    mut dust_coins_total: u64,
+) -> impl Iterator<Item = CoinsToSpendIndexEntry> {
+    big_coins.into_iter().skip_while(move |item| {
+        let amount = item.0.amount();
+        dust_coins_total
+            .checked_sub(amount)
+            .map(|new_value| {
+                dust_coins_total = new_value;
```

Comment:
```suggestion
fn skip_big_coins_up_to_amount<'a>(
    big_coins: impl IntoIterator<Item = CoinsToSpendIndexEntry> + 'a,
    dust_coins_total: &'a mut u64,
) -> impl Iterator<Item = CoinsToSpendIndexEntry> + 'a {
    big_coins.into_iter().skip_while(move |item| {
        let amount = item.0.amount();
        dust_coins_total
            .checked_sub(amount)
            .map(|new_value| {
                *dust_coins_total = new_value;
```

Can we use reference here?=D I don't trust closures with `mut` and `move` =D Sorry for being paranoic=)

Reply:
Actually, our intention here is _not_ to modify the value outside of this function (on the callsite). The `mut` in `mut dust_coins_total: u64,` is just a shortcut allowing to use this binding as a counter meaningful only within the closure.

I modified the code to show the intent more explicitly in this commit: c001fd32613b4258ec119ee2ebe1722a629da925

Reply:
I was not sure that it works, because I don't know where `dust_coins_total` lives while iterator is alive. But seems it is "Rust's magic"=) Because it works in playground

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a01bef2288c09f61ac6691bff996144a

<img width="1728" alt="image" src="https://github.com/user-attachments/assets/4fe24d0e-f3bc-4e17-add1-745ed4bce5cc" />